### PR TITLE
bullet constraints

### PR DESCRIPTION
**Pull request by sethalves**
- add framework for exposing bullet constraints to javascript
- add hinge-vs-world constraint
- add userData flag to indicate that an object should not be forced kinematic when near-grabbed

### CODE REVIEW

**Comment by sethalves** [comment]
to test:
- make sure far-grab and grab.js work on dynamic entities
- run this in the console:
```javascript
var pos = Vec3.sum(MyAvatar.position, Vec3.multiplyQbyV(MyAvatar.orientation, {x: 0, y: 0.1, z: -2}));

var doorID = Entities.addEntity({
    name: "door text",
    type: "Box",
    color: { blue: 128, green: 20, red: 20 },
    dimensions: { x: 1.0, y: 2, z: 0.1 },
    position: pos,
    dynamic: true,
    collisionless: false,
    gravity: { x: 0, y: 0, z: 0 },
    lifetime: 6000
});

var hingeID = Entities.addAction("hinge", doorID, {
    pivot: { x: -0.5, y: 0, z: 0 },
    axis: { x: 0, y: 1, z: 0 },
    tag: "door hinge test",
    ttl: 6000
});
```

**Comment by hifi-gustavo** [comment]
Build console output for commit: 507f658a5f0f858552c06b6569b5e88567eb1870 available at: https://gist.github.com/bde067fb6bbb88a2a0074465298b0823<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10201-507f658.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10201-507f658.dmg

**Comment by Nex-Pro** [comment]
### PASS
Was able to create hinged entities. 
Also by modifying the script change the hinge axis and location.

Nice work.  

Note:

With the hinge point at the top of the entity gravity is currently ignored and ents rotating around an upper hinge point do not settle with the mass down.  I look forward to seeing this plugged into the edit system.
As expected any edit of the hinged ent results in the hinge failing.

**Comment by sethalves** [comment]
Hm yeah it looks like edit.js will have to remove or disable a hinge before it can change the location of an entity, and then put it back.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 026dde2e6873afd88ef51701a8e259e41f269374 available at: https://gist.github.com/bde067fb6bbb88a2a0074465298b0823<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10201-026dde2.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10201-026dde2.dmg

File: libraries/entities/src/EntityDynamicInterface.cpp [hunk 1]
```diff
@@ -0,0 +1,345 @@
...
+    return stream << (quint16)entityDynamicType;
+}
+
+QDataStream& operator>>(QDataStream& stream, EntityDynamicType& entityDynamicType)
+{
+    quint16 dynamicTypeAsInt;
+    stream >> dynamicTypeAsInt;
+    entityDynamicType = (EntityDynamicType)dynamicTypeAsInt;
+    return stream;
+}
```
Comment (AndrewMeadows) [inline libraries/entities/src/EntityDynamicInterface.cpp]: Bad brace formatting here and the function above.

Comment (sethalves) [inline libraries/entities/src/EntityDynamicInterface.cpp]: fixed

**Comment by hifi-gustavo** [comment]
Build console output for commit: 9b12956909bf2ccedea917fc17aaaaba5f790142 available at: https://gist.github.com/bde067fb6bbb88a2a0074465298b0823<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10201-9b12956.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10201-9b12956.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 1ab3a1eb60986f4ac058d3bba2b0ff80b7b0b7d0 available at: https://gist.github.com/bde067fb6bbb88a2a0074465298b0823<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10201-1ab3a1e.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10201-1ab3a1e.dmg

**Comment by sethalves** [comment]
Closing this one in favour of PR I'm about to open...

**Pull request closed without merging**